module/os/linux/zfs/vdev_disk.c

```diff
@@ -670,126 +591,475 @@ vdev_bio_alloc(struct block_device *bdev, gfp_t gfp_mask,
+
+	return (s.npages);
+}
 
-		/* Matching put called by vdev_disk_physio_completion */
-		vdev_disk_dio_get(dr);
+/* Iterator callback to submit ABD pages to the vbio. */
+static int
+vdev_disk_fill_vbio_cb(void *buf, size_t len, void *priv)
+{
+	vbio_t *vbio = priv;
+	int err;
+
+	struct page *page;
+	uint_t off;
+
+	while (len > 0) {
+		_buf_to_page_and_offset(buf, &page, &off);
 
-		BIO_BI_SECTOR(dr->dr_bio[i]) = bio_offset >> 9;
-		dr->dr_bio[i]->bi_end_io = vdev_disk_physio_completion;
-		dr->dr_bio[i]->bi_private = dr;
-		bio_set_op_attrs(dr->dr_bio[i], rw, flags);
+		uint_t take = MIN(len, page_size(page)-off);
 
-		/* Remaining size is returned to become the new size */
-		bio_size = abd_bio_map_off(dr->dr_bio[i], zio->io_abd,
-		    bio_size, abd_offset);
+		err = vbio_add_page(vbio, page, take, off);
+		if (err != 0)
+			return (err);
 
-		/* Advance in buffer and construct another bio if needed */
-		abd_offset += BIO_BI_SIZE(dr->dr_bio[i]);
-		bio_offset += BIO_BI_SIZE(dr->dr_bio[i]);
+		buf += take;
+		len -= take;
 	}
 
-	/* Extra reference to protect dio_request during vdev_submit_bio */
-	vdev_disk_dio_get(dr);
+	return (0);
+}
 
-	if (dr->dr_bio_count > 1)
-		blk_start_plug(&plug);
+static int
+vdev_disk_io_rw(zio_t *zio)
+{
+	vdev_t *v = zio->io_vd;
+	vdev_disk_t *vd = v->vdev_tsd;
+	struct block_device *bdev = vd->vd_bdev;
+	int flags = 0;
+
+	/*
+	 * Accessing outside the block device is never allowed.
+	 */
+	if (zio->io_offset + zio->io_size > bdev->bd_inode->i_size) {
+		vdev_dbgmsg(zio->io_vd,
+		    "Illegal access %llu size %llu, device size %llu",
+		    (u_longlong_t)io_offset,
+		    (u_longlong_t)io_size,
+		    (u_longlong_t)i_size_read(bdev->bd_inode));
+		return (SET_ERROR(EIO));
+	}
 
-	/* Submit all bio's associated with this dio */
-	for (int i = 0; i < dr->dr_bio_count; i++) {
-		if (dr->dr_bio[i])
-			vdev_submit_bio(dr->dr_bio[i]);
+	if (!(zio->io_flags & (ZIO_FLAG_IO_RETRY | ZIO_FLAG_TRYHARD)) &&
+	    v->vdev_failfast == B_TRUE) {
+		bio_set_flags_failfast(bdev, &flags, zfs_vdev_failfast_mask & 1,
+		    zfs_vdev_failfast_mask & 2, zfs_vdev_failfast_mask & 4);
 	}
 
-	if (dr->dr_bio_count > 1)
-		blk_finish_plug(&plug);
+	/*
+	 * Count the number of pages in the incoming ABD, and check its
+	 * alignment. If any part of it would require submitting a page that is
+	 * not aligned to the logical block size, then we take a copy into a
+	 * linear buffer and submit that instead. This should be impossible on
+	 * a 512b LBS, and fairly rare on 4K, usually requiring
+	 * abnormally-small data blocks (eg gang blocks) mixed into the same
+	 * ABD as larger ones (eg aggregated).
+	 */
+
+	/* LBS is guaranteed to be one of 512, 1024, 2048 or 4096 */
+	uint_t lbs = bdev_logical_block_size(bdev);
 
-	vdev_disk_dio_put(dr);
+	abd_t *abd = zio->io_abd;
+	uint_t npages = vdev_disk_check_pages(abd, zio->io_size, lbs);
+	if (npages == 0) {
+		void *buf;
+		/* We can't safely use pages from this ABD, so use a copy */
```

Comment:
Nice, I wanted the same thing but didn't really know where to put it. Will do!

Reply:
Added `borrow_copy_total` and `borrow_copy_size_total`. The `*_cnt` stats are counting what's alive right now, but borrows are too transient to usefully count them that way. Also only counting copies; just borrowing isn't really that big a deal for linear abds.

This isn't the most useful stat perhaps, because there's other places that borrows happen, but it should be ok for order-of-magnitude comparisons. Probably useful to add stats to the vdev itself sometime, but I won't do that right now.